java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/SurroundWithHint.java

```diff
@@ -0,0 +1,298 @@
+        if (info == null) {
+            return Collections.emptyList();
+        }
+        info.toPhase(JavaSource.Phase.ELEMENTS_RESOLVED);
+        int startOffset = getOffset(info, params.getRange().getStart());
+        int endOffset = getOffset(info, params.getRange().getEnd());
+        if (startOffset >= endOffset) {
+            return Collections.emptyList();
+        }
+        Document doc;
+        try {
+            doc = info.getDocument();
+        } catch (IOException ex) {
+            return Collections.emptyList();
+        }
+        Tree.Kind treeKindCtx = null;
+        String stringCtx = null;
+        TreeUtilities tu = info.getTreeUtilities();
+        TokenSequence<JavaTokenId> ts = SourceUtils.getJavaTokenSequence(info.getTokenHierarchy(), startOffset);
+        int delta = ts.move(startOffset);
+        if (delta == 0 || ts.moveNext() && ts.token().id() == JavaTokenId.WHITESPACE) {
+            delta = ts.move(endOffset);
+            if (delta == 0 || ts.moveNext() && ts.token().id() == JavaTokenId.WHITESPACE) {
+                String selectedText = info.getText().substring(startOffset, endOffset).trim();
+                SourcePositions[] sp = new SourcePositions[1];
+                ExpressionTree expr = selectedText.length() > 0 ? tu.parseExpression(selectedText, sp) : null;
+                if (expr != null && expr.getKind() != Tree.Kind.IDENTIFIER && !Utilities.containErrors(expr) && sp[0].getEndPosition(null, expr) >= selectedText.length()) {
+                    stringCtx = EXPRESSION;
+                }
+            }
+        }
+        Tree tree = tu.pathFor(startOffset).getLeaf();
+        if (tu.pathFor(endOffset).getLeaf() == tree) {
+            treeKindCtx = tree.getKind();
+            switch (treeKindCtx) {
+                case CASE:
+                    if (startOffset < info.getTrees().getSourcePositions().getEndPosition(info.getCompilationUnit(), ((CaseTree)tree).getExpression())) {
+                        treeKindCtx = null;
+                    }
+                    break;
+                case CLASS:
+                    SourcePositions sp = info.getTrees().getSourcePositions();
+                    int startPos = (int)sp.getEndPosition(info.getCompilationUnit(), ((ClassTree)tree).getModifiers());
+                    if (startPos <= 0) {
+                        startPos = (int)sp.getStartPosition(info.getCompilationUnit(), tree);
+                    }
+                    String headerText = info.getText().substring(startPos, startOffset);
+                    int idx = headerText.indexOf('{'); //NOI18N
+                    if (idx < 0) {
+                        treeKindCtx = null;
+                        stringCtx = CLASS_HEADER;
+                    }
+                    break;
+                case FOR_LOOP:
+                case ENHANCED_FOR_LOOP:
+                    if (!isRightParenthesisOfLoopPresent(info, startOffset)) {
```

Comment:
Maybe missing a testcase for this TreeKind...

Reply:
Replaced with a call to the existing `CodeTemplateFilter`s.